### PR TITLE
feat(data-protection-complaint): Conditionally required input implemented

### PR DESCRIPTION
**Pull request by Thorrik58**
# Conditionally required input implemented

Attach a link to issue if relevant

## What

Conditionally required input implemented

## Why

Personuvernd ran into an bug where a user could select other reason without specifying.

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/application/templates/data-protection-complaint/src/forms/ComplaintForm.ts [hunk 1]
```diff
@@ -517,72 +517,10 @@ export const ComplaintForm: Form = buildForm({
...
-                  id: 'subjectOfComplaint.somethingElse',
-                  title: complaint.labels.subjectSomethingElse,
-                  placeholder: complaint.labels.subjectSomethingElsePlaceholder,
-                  backgroundColor: 'blue',
-                  condition: (formValue) => {
-                    const values =
-                      ((formValue.subjectOfComplaint as FormValue)
-                        ?.values as string[]) || []
-                    return values.includes('other')
-                  },
```
[the remark this answers is not in the document]
Comment (Thorrik58) [reply libs/application/templates/data-protection-complaint/src/forms/ComplaintForm.ts:585]: Agreed! This is way too much hassle.

File: libs/application/templates/data-protection-complaint/src/fields/ConditionallyRequiredInput.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+import { complaint } from '../lib/messages'
+
+export const ReasonsForComplaint: FC<FieldBaseProps> = ({
+  field,
+  application,
+  error,
+}) => {
+  const { register, errors, setValue } = useFormContext()
+  const [statefulOtherReason, setStatefulOtherReason] = useState(
+    getValueViaPath(application.answers, 'otherReason' || '') as string,
```
[the remark this answers is not in the document]
Comment (Thorrik58) [reply libs/application/templates/data-protection-complaint/src/fields/ConditionallyRequiredInput.tsx:25]: SomethingElse is just for saving the input in an optional field where the user can fill in the reason. 
The SomethingElseValue is the hidden value which is enforced by the dataschema to be always required. 

To get the optionally required functionality working I needed to mix the two together and assign a fallback value to the SomethingElseValue when the "other" checkbox wasnt checked - therefore making the value optional.
image

Once the user checks that checkbox the value of SomethingElseValue gets cleared and the user is required to fill in the value in SomethingElse in order for it to get assigned to SomethingElseValue to fulfil the dataschema.

Hope this explanation is clear enough.

[the remark this answers is not in the document]
Comment (eirikurn) [reply libs/application/templates/data-protection-complaint/src/fields/ConditionallyRequiredInput.tsx:25]: I'm sorry for your pain. Hope we can improve this soon.

File: libs/application/templates/data-protection-complaint/src/fields/ConditionallyRequiredInput.tsx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
+  ) => {
+    setStatefulOtherReason(e.target.value)
+  }
+
+  const getValueConditionallyRequiredValue = () => {
+    if (statefulCheckbox.includes('other')) {
+      return statefulOtherReason
+    }
+    return 'someValue'
```
[the remark this answers is not in the document]
Comment (eirikurn) [reply libs/application/templates/data-protection-complaint/src/fields/ConditionallyRequiredInput.tsx:52]: No, I'm not familiar enough with the validation system. I just noticed you pass `required: true` when registering the optional field. Was just wondering if you could render the actual input with `display={none} required={false}` at runtime, but I guess you tried that already.

**Review by eirikurn — APPROVED** [review]

**Review by obmagnusson — APPROVED** [review]

**Review by sigruntg — APPROVED** [review]

**Pull request merged**